Codequality

CSchoel opened the pull request:
I created this pull request prematurely in order to be able to provide a comprehensive code review. This branch is still work in progress.

CptKaNe committed:
made launchclient static again to prevent the usage of the static field executor from a non static method

CptKaNe committed:
added a short comment to not yet implemented methods

CptKaNe committed:
implemented a static method setExecutorPool() and call it in the main on both client and serverside

CptKaNe committed:
added a default case for the switch in MopeWorkspaceService initialized fields called in consoleclient and server constructor statically

CSchoel commented on src/main/java/Client/ConsoleClientLauncher.java:
```diff
@@ -28,17 +27,25 @@
     private static final Logger logger = LoggerFactory.getLogger(ConsoleClientLauncher.class);
     public static Future<Void> clientListening;
 
+    static {
+        socket = null;
+        client = null;
+        menu = null;
+    }
+
```
This is not the solution you are looking for. :hand: 

Static blocks are only seldom useful (as "constructors" for static classes, for example).

The real issue here is that the *variables* are static, but they are *assigned* in a non-static context (that is the constructor). You should think about which makes more sense: Having multiple `ConsoleClientLauncher` objects with different values for these variables, or having *one* single ConsoleClientLauncher that uses only static methods. (Or option 3: Use the singleton pattern.)

CptKaNe replied:
You are absolutely right, I was insecure about supporting more clients at the same time, so i was trying to find a proper solution. Because if there is only one Launcher with static methods, multiclients are not possible or am i mixing things up in this case?

CSchoel commented on src/main/java/Client/ConsoleClientLauncher.java:
```diff
@@ -28,17 +27,25 @@
         socket = new Socket(host, port);
         menu = new ConsoleMenu(client);
     }
+    public static void setExecutorPool() {
+        executor = Executors.newFixedThreadPool(2);
+    }
```
I am confused why this is required by SonarQube. This looks like an unnecessary method that could be replaced by the one line of code that it contains.

CSchoel replied:
@CptKaNe Can you tell me what the error is that you tried to fix here?

CptKaNe replied:
yes, it also was referring of a static field (executor) in a non static context, but i guess you already find out because of the further feedback below. Lots of work to do for me ;-)

CSchoel commented on src/main/java/Server/MopeLSPServerLauncher.java:
```diff
@@ -29,20 +29,29 @@
+        server = null;
+        configObject = null;
+    }
+
```
Again, this is not what you want here. Decide how the class should be used: As instance, as static class, or as singleton, and refactor it accordingly.

CSchoel commented on src/main/java/Server/MopeLSPServerLauncher.java:
```diff
@@ -29,20 +29,29 @@
         logger.info("Server socket listening on port " + configObject.port );
         System.out.flush();
-        executor = Executors.newCachedThreadPool();
+        //executor = Executors.newCachedThreadPool();
```
There is no need to keep deleted code as comments in a git repository. That's what version control is for.

CSchoel replied:
Also, creating the executor in `main()` instead of `launchServer()` means that the tests now fail with a null pointer exception, since the `ServerIntegrationTest` calls `launchServer()` but not `main()`.

CptKaNe replied:
Well this is probably because i was not into the project for weeks and do not know every change since then, but yesterday I already had the feeling, theres something wrong with these changes. I will appreciate it.

CSchoel commented on src/main/java/Server/MopeWorkspaceService.java:
```diff
@@ -55,6 +57,9 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams didChangeWatchedFi
                     result = modelicaService.getCompilerVersion();
                     break;
+                default:
+                    result = CompletableFuture.completedFuture(error);
```
Is this really what you want? It is a design choice, but inside a Java environment I would opt for a more rigid approach: When you get an unexpected input, you throw a `InputMismatchException`.

CptKaNe replied:
Thanks for the hint!

manuEbg commented:
I canceled the workflow because testing already took 1 hour and 50 Minutes. So I think there is something wrong...
Maybe @CptKaNe  should have a look at this

CSchoel reviewed (commented):
I went through the changes and provided feedback as detailed as possible. There are a few misconceptions here that should be addressed.

CSchoel commented:
> I canceled the workflow because testing already took 1 hour and 50 Minutes. So I think there is something wrong...
> Maybe @CptKaNe should have a look at this

I have found the culprit: It is actually a `NullPointerException`, because the `executor` is not initialized for the `ServerIntegrationTest`. This raises another issue: Why does GH actions not pick up this error as it happens. I know there is an issue with GH action not recognizing process exits correctly, but I did not think this would be the case for a simple Gradle task.

CSchoel commented:
Switched target branch to dev, so that we can actually use this pull request.

CptKaNe committed:
- rolled back some bad changes concerning static initialization
- added exceptions for workspaceservice
- updated syntax of commands, now start with lower letter
- Tried to implement the singleton pattern, but removed this changes because they were wrong
- ignored some unneccessary code smells from sonarqube
